### Loading Dataset
Load the retail sales dataset into Pandas Dataframe for inspection and analysis

In [2]:
import pandas as pd
import os
df=pd.read_csv("../data/Sample_ Superstore.csv")


### Initial Data Inspection

Perform an initial inspection to understand the dataset structure before cleaning.


In [3]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Segment,Country,City,State,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11-08-2016,11-11-2016,Second Class,CG-12520,Consumer,United States,Henderson,Kentucky,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11-08-2016,11-11-2016,Second Class,CG-12520,Consumer,United States,Henderson,Kentucky,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,06-12-2016,6/16/2016,Second Class,DV-13045,Corporate,United States,Los Angeles,California,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10-11-2015,10/18/2015,Standard Class,SO-20335,Consumer,United States,Fort Lauderdale,Florida,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10-11-2015,10/18/2015,Standard Class,SO-20335,Consumer,United States,Fort Lauderdale,Florida,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [4]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Segment', 'Country', 'City', 'State', 'Region',
       'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Sales',
       'Quantity', 'Discount', 'Profit'],
      dtype='str')

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 19 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Row ID        9994 non-null   int64  
 1   Order ID      9994 non-null   str    
 2   Order Date    9994 non-null   str    
 3   Ship Date     9994 non-null   str    
 4   Ship Mode     9994 non-null   str    
 5   Customer ID   9994 non-null   str    
 6   Segment       9994 non-null   str    
 7   Country       9994 non-null   str    
 8   City          9994 non-null   str    
 9   State         9994 non-null   str    
 10  Region        9994 non-null   str    
 11  Product ID    9994 non-null   str    
 12  Category      9994 non-null   str    
 13  Sub-Category  9994 non-null   str    
 14  Product Name  9994 non-null   str    
 15  Sales         9994 non-null   float64
 16  Quantity      9994 non-null   int64  
 17  Discount      9994 non-null   float64
 18  Profit        9994 non-null   float64
d

In [6]:
df.describe()

,Row ID,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,229.858001,3.789574,0.156203,28.656896
std,2885.163629,623.245101,2.225110,0.206452,234.260108
min,1.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,22638.480000,14.000000,0.800000,8399.976000


In [7]:
df.shape

(9994, 19)

### Dataset Overview
- Rows: 9994 
- Columns: 19
- Numerical columns: 5 and Categorical columns: 14
- Order Date and Ship Date are stored as object type and require conversion to datetime before time-based analysis



### Key Columns
- **Sales**: Revenue generated from each product sale
- **Profit**: Profit earned on each transaction after subtracting business costs. Can be positive or negative
- **Discount**: Discount applied to the sale
- **Quantity**: Number of units sold

### Initial Business Observation
- The dataset contains transaction-level retail sales records.
- Each row represents a product purchased within an order.
- Order ID is not unique because one order can contain multiple products.


### Data Quality Assessment


In [8]:
df.isna().sum()

Row ID          0
Order ID        0
Order Date      0
Ship Date       0
Ship Mode       0
Customer ID     0
Segment         0
Country         0
City            0
State           0
Region          0
Product ID      0
Category        0
Sub-Category    0
Product Name    0
Sales           0
Quantity        0
Discount        0
Profit          0
dtype: int64

In [9]:
df.duplicated().sum()

np.int64(0)

**Observation**
- No missing values were found in any column.
- No duplicate rows were found in dataset.

In [10]:
df['Order Date']=df['Order Date'].str.replace('-','/', regex=True)

In [11]:
df['Order Date']=pd.to_datetime(df['Order Date'],format='%m/%d/%Y')


In [12]:
df['Ship Date']=df['Ship Date'].str.replace('-','/',regex=True)

In [13]:
df['Ship Date']=pd.to_datetime(df['Ship Date'],format='%m/%d/%Y')

**Observation**
- 'Order Date' and 'Ship Date' were stored as the 'object' data types.
- The date seperator was inconsistent.
- The seperators were standardized to 'datetime' datatype.


In [39]:
df.sample(5)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Segment,Country,City,State,...,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Year,Month,Day
4233,4234,CA-2017-100223,2017-07-05,2017-07-10,Standard Class,LS-16945,Corporate,United States,Dallas,Texas,...,Office Supplies,Binders,Tuf-Vin Binders,6.316,1,0.8,-10.4214,2017,7,5
7882,7883,CA-2017-118017,2017-12-03,2017-12-06,Second Class,LC-16870,Consumer,United States,Thornton,Colorado,...,Technology,Accessories,Memorex Micro Travel Drive 16 GB,102.336,8,0.2,14.0712,2017,12,3
8838,8839,CA-2014-153808,2014-04-06,2014-04-10,Second Class,FH-14275,Corporate,United States,Los Angeles,California,...,Office Supplies,Art,Boston Home & Office Model 2000 Electric Penci...,70.950,3,0.0,18.4470,2014,4,6
5853,5854,US-2014-138835,2014-09-10,2014-09-14,Standard Class,DA-13450,Home Office,United States,Virginia Beach,Virginia,...,Office Supplies,Binders,Tuff Stuff Recycled Round Ring Binders,9.640,2,0.0,4.7236,2014,9,10
4122,4123,CA-2017-133823,2017-05-11,2017-05-12,Same Day,LP-17080,Consumer,United States,Seattle,Washington,...,Office Supplies,Labels,Avery 50,37.590,3,0.0,17.6673,2017,5,11


In [15]:
df['Sales'].describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Sales, dtype: float64

In [16]:
df['Discount'].describe()

count    9994.000000
mean        0.156203
std         0.206452
min         0.000000
25%         0.000000
50%         0.200000
75%         0.200000
max         0.800000
Name: Discount, dtype: float64

In [17]:
df['Quantity'].describe()

count    9994.000000
mean        3.789574
std         2.225110
min         1.000000
25%         2.000000
50%         3.000000
75%         5.000000
max        14.000000
Name: Quantity, dtype: float64

In [18]:
df['Profit'].describe()

count    9994.000000
mean       28.656896
std       234.260108
min     -6599.978000
25%         1.728750
50%         8.666500
75%        29.364000
max      8399.976000
Name: Profit, dtype: float64

**Observation**
- Sales, Quantity, Discount contains no invalid negative values.
- The mean Sales (229.85) is considerably higher than median(54.49).
- The mean Profit(28.65) is considered higher than median(8.67).
- This both suggest the sales and profit distribution is right-skewed, indicating possible presence of high-value outliers.
- Negative Profit values are present and represent loss-making transactions rather than data quality issues.


In [19]:
df['Region'].unique()

<StringArray>
['South', 'West', 'Central', 'East']
Length: 4, dtype: str

In [20]:
df['Country'].unique()

<StringArray>
['United States']
Length: 1, dtype: str

In [21]:
df['State'].unique()

<StringArray>
[            'Kentucky',           'California',              'Florida',
       'North Carolina',           'Washington',                'Texas',
            'Wisconsin',                 'Utah',             'Nebraska',
         'Pennsylvania',             'Illinois',            'Minnesota',
             'Michigan',             'Delaware',              'Indiana',
             'New York',              'Arizona',             'Virginia',
            'Tennessee',              'Alabama',       'South Carolina',
               'Oregon',             'Colorado',                 'Iowa',
                 'Ohio',             'Missouri',             'Oklahoma',
           'New Mexico',            'Louisiana',          'Connecticut',
           'New Jersey',        'Massachusetts',              'Georgia',
               'Nevada',         'Rhode Island',          'Mississippi',
             'Arkansas',              'Montana',        'New Hampshire',
             'Maryland', 'District of

In [22]:
df['Segment'].unique()

<StringArray>
['Consumer', 'Corporate', 'Home Office']
Length: 3, dtype: str

In [23]:
df['Category'].unique()

<StringArray>
['Furniture', 'Office Supplies', 'Technology']
Length: 3, dtype: str

In [24]:
df['Sub-Category'].unique()

<StringArray>
[  'Bookcases',      'Chairs',      'Labels',      'Tables',     'Storage',
 'Furnishings',         'Art',      'Phones',     'Binders',  'Appliances',
       'Paper', 'Accessories',   'Envelopes',   'Fasteners',    'Supplies',
    'Machines',     'Copiers']
Length: 17, dtype: str

In [25]:
df['Ship Mode'].unique()

<StringArray>
['Second Class', 'Standard Class', 'First Class', 'Same Day']
Length: 4, dtype: str

In [26]:

pd.Series(df['City'].sort_values(ascending=True).unique())


0         Aberdeen
1          Abilene
2            Akron
3      Albuquerque
4       Alexandria
          ...     
526     Woonsocket
527        Yonkers
528           York
529        Yucaipa
530           Yuma
Length: 531, dtype: str

In [27]:
df['City'].str.capitalize()

0             Henderson
1             Henderson
2           Los angeles
3       Fort lauderdale
4       Fort lauderdale
             ...       
9989              Miami
9990         Costa mesa
9991         Costa mesa
9992         Costa mesa
9993        Westminster
Name: City, Length: 9994, dtype: str

In [28]:
df['City'].str.strip()

0             Henderson
1             Henderson
2           Los Angeles
3       Fort Lauderdale
4       Fort Lauderdale
             ...       
9989              Miami
9990         Costa Mesa
9991         Costa Mesa
9992         Costa Mesa
9993        Westminster
Name: City, Length: 9994, dtype: str

**Observation**
- No missing values, inconcistent capitalization, leading/trailing spaces or obvious invalid values were identified.
- No additional cleaning was required for the categorical columns.

In [42]:
df.sample()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Segment,Country,City,State,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
1994,1995,CA-2017-133648,2017-06-25,2017-07-02,Standard Class,ML-17755,Home Office,United States,Springfield,Oregon,West,OFF-FA-10004395,Office Supplies,Fasteners,Plymouth Boxed Rubber Bands by Plymouth,11.304,3,0.2,-2.1195


In [44]:
df.to_csv('../data/Sample_Superstore_Cleaned.csv',index=False)

In [46]:
df=pd.read_csv('../data/Sample_Superstore_Cleaned.csv')
print(df.shape)


(9994, 19)


In [50]:
df.iloc[150]

Row ID                                       151
Order ID                          CA-2016-114489
Order Date                            2016-12-05
Ship Date                             2016-12-09
Ship Mode                         Standard Class
Customer ID                             JE-16165
Segment                                Corporate
Country                            United States
City                                    Franklin
State                                  Wisconsin
Region                                   Central
Product ID                       OFF-BI-10002735
Category                         Office Supplies
Sub-Category                             Binders
Product Name    GBC Prestige Therm-A-Bind Covers
Sales                                     171.55
Quantity                                       5
Discount                                     0.0
Profit                                   80.6285
Name: 150, dtype: object